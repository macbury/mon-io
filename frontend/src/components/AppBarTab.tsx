import React from 'react'
import { NavigationNavigateAction } from 'react-navigation'
import styled from 'styled-components/native'
import Link from './Link'
import Icon from './Icon'

interface IButtonInnerProps {
  current: boolean
}

const LinkContainer = styled(Link)<IButtonInnerProps>`
  display: flex;
  justify-content: center;
  padding-top: 3px;
  height: 50px;
  padding-bottom: ${({ current }) => current ? 0 : 3};
  border-bottom-width: ${({ current }) => current ? 3 : 0};
  border-bottom-color: ${(props) => props.theme.headerItemBottomColor};
`

const ButtonInner = styled.View<IButtonInnerProps>`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  opacity: ${({ current }) => current ? 1.0 : 0.6};
  padding: 0px 15px;
`

export interface IAppBarTab {
  icon: string
  current?: boolean
  action: NavigationNavigateAction
}

export default function AppBarTab(props: IAppBarTab) {
  const { icon, current, action }  = props
  return (
    <LinkContainer action={action} ripple current={current}>
      <ButtonInner current={current}>
        <Icon name={icon} size={28} />
      </ButtonInner>
    </LinkContainer>
  )
}
