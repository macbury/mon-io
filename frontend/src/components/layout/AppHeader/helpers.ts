import { AppHeaderProp } from './types'

export function getTitleFromNavigation(props : AppHeaderProp) : string {
  let title = props.title
  if (!title) {
    title = props.scene?.descriptor?.options?.title
  }
  return title
}