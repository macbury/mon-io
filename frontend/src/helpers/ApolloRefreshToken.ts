import SessionStore from '../stores/SessionStore'

export default class ApolloRefreshToken {
  private refreshing : boolean;
  private resolvers : Array<Function> = []
  private sessions : SessionStore

  constructor(sessions : SessionStore) {
    this.sessions = sessions
  }

  public refreshToken() : Promise<void> {
    return new Promise<void>((resolve) => {
      this.resolvers.push(resolve)

      if (!this.refreshing) {
        this.sessions.refreshServerToken().then(this.onRefreshEnd).catch(this.onRefreshEnd)
        this.refreshing = true
      }
    })
  }

  private onRefreshEnd = () => {
    this.refreshing = false
    this.resolvers.forEach((resolve) => resolve())
    this.resolvers = []
  }
}