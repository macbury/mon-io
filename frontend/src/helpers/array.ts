export function chunk(arr : Array<any>, len : number) {
  var chunks = [],
      i = 0,
      n = arr.length;

  while (i < n) {
    chunks.push(arr.slice(i, i += len));
  }

  return chunks;
}

export function chunks(arr : Array<any>, lens : number[]) {
  var chunks = [],
      i = 0,
      n = arr.length;
  let len = lens.shift()

  while (i < n) {
    chunks.push(arr.slice(i, i += len));
    if (lens.length > 0) {
      len = lens.shift()
    }
  }

  return chunks;
}

