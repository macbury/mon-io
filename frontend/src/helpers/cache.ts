
//import { Asset } from 'react-native'
import { Image } from 'react-native'
//import * as Font from 'expo-font'

export function cacheImages(images : Array<string>) {
  return images.map(image => (Image.prefetch(image)));
}

export function cacheFonts(fonts : Array<any>) {
  return [] //fonts.map(font => Font.loadAsync(font));
}