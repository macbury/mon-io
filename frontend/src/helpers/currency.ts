import { useMemo } from 'react'
import { Money } from '../api/graphql'

export function formatMoney(money : Money, abs = false, round = false, minimumFractionDigits?: number) {
  if (!money) {
    return '-'
  }

  const amount = abs ? Math.abs(money.cents) : money.cents

  if (!minimumFractionDigits) {
    minimumFractionDigits = money.currency.subunitToUnit.toString().split('').length - 1
  }

  if (round) {
    return Math.round(amount / money.currency.subunitToUnit).toLocaleString(undefined, { minimumFractionDigits }) + ' ' + money.currency.symbol
  } else {
    return (amount / money.currency.subunitToUnit).toLocaleString(undefined, { minimumFractionDigits }) + ' ' + money.currency.symbol
  }
}

export function moneyParts(money : Money) {
  if (money && money.currency) {
    let [dollars, cents] = (money.cents / money.currency.subunitToUnit).toPrecision().split('.')
    if (!cents) {
      cents = '00'
    }
    if (cents.length == 1) {
      cents = `${cents}0`
    }
    return [parseInt(dollars), cents, money.currency.symbol]
  } else {
    return [0, 0, 'EUR']
  }
}

export function useFormatMoney(money : Money, abs = false, round = false, minimumFractionDigits?: number) {
  return useMemo(() => formatMoney(money, abs, round, minimumFractionDigits), [money, abs, round, minimumFractionDigits])
}

export function useMoneyParts(money : Money) {
  return useMemo(() => moneyParts(money), [money])
}