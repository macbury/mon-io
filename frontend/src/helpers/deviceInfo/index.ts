import DeviceInfo from 'react-native-device-info'

export function getDeviceName() : Promise<String> {
  return DeviceInfo.getDeviceName()
}