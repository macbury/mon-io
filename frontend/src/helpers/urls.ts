import { Recurrence } from './../api/graphql';
import moment, { Moment } from 'moment-timezone'
import { NavigationActions, NavigationNavigateAction } from 'react-navigation'
import { TransactionCategoryKind, RecurrenceUpdateMode } from '../api/graphql'

export function homePath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'home' })
}

export function receiptsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'receipts' })
}

export function mapPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'map' })
}

export function authPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function importsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'imports' })
}

export function editImportPath(importId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newImport', params: { importId } })
}

export function settingsApiExplorerPath(refreshTokenId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'apiExplorer', params: { refreshTokenId } })
}

export function categoriesPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'categories' })
}

export function newCategoryPath(kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newCategory', params: { kind } })
}

export function editCategoriesPath(categoryId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategory', params: { categoryId } })
}

export function scanPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'scan' })
}

export function changeNewTransactionCategoryPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'changeNewTransactionCategory' })
}

export function signInPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function seriesPath(date?: Moment) : NavigationNavigateAction {
  const params = date ? { date: date.format('YYYY-MM-DD') } : {}
  return NavigationActions.navigate({ routeName: 'series', params })
}

export function chartsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'charts' })
}

export function editSeriesPath(seriesId : string, updateType: RecurrenceUpdateMode, occuredAt?: Moment | string) : NavigationNavigateAction {
  let date = occuredAt ? moment(occuredAt).format('YYYY-MM-DD') : null
  return NavigationActions.navigate({ routeName: 'editSeries', params: { seriesId, date, updateType } })
}

export function editCategoryBudgetPath(budgetId : string, categoryId : string, kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategoryBudget', params: { budgetId, categoryId, kind } })
}

export function budgetPath(date : Moment) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'budget', params: { date: date.format('YYYY-MM') } })
}

export function summaryPath(date : Moment, type?: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'summary', params: { date: date.format('YYYY-MM'), type } })
}

export function transactionsPath(date? : Moment, type?: TransactionCategoryKind, importId?: string) : NavigationNavigateAction {
  return NavigationActions.navigate({
    routeName: 'transactions',
    params: {
      date: date?.format('YYYY-MM'),
      type,
      importId
    }
  })
}

export function previewReceiptPath(receiptId : string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'preview', params: { receiptId } })
}

export function settingsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settings' })
}

export function settingsRefreshTokensPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsRefreshTokens' })
}

export function settingsAuthorizeDevicePath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsQuickLogin' })
}

export function settingsDownloadMobileAppPath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsDownloadMobileApp' })
}

export function newSummaryTransactionPath(categoryId: string, createdAt: string, kind: TransactionCategoryKind) {
  return NavigationActions.navigate({
    routeName: 'newSummaryTransaction',
    params: {
      categoryId, createdAt, kind
    }
  })
}

export function newBillTransactionPath(receiptId: string) {
  return NavigationActions.navigate({ routeName: 'newBillTransaction', params: { receiptId } })
}

export function newPlannedTransactionPath(seriesId: string, createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newPlannedTransaction', params: { createdAt, seriesId } })
}

export function newSeriesPath(createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newSeries', params: { createdAt, recurrence: Recurrence.Once } })
}

export function editSummaryTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editSummaryTransaction', params: { transactionId } })
}

export function editTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editTransaction', params: { transactionId } })
}