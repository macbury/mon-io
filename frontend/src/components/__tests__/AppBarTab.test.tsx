import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import AppBarTab, { IAppBarTab } from '../AppBarTab'
import { homePath } from '../../helpers/urls'

it('renders correctly', () => {
  const props : IAppBarTab = {
    action: homePath(),
    icon: 'repeat',
    current: false
  }

  const component = shallow(<AppBarTab {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})