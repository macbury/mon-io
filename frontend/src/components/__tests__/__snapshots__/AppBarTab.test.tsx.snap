// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Styled(Link)
  action={
    Object {
      "routeName": "home",
      "type": "Navigation/NAVIGATE",
    }
  }
  current={false}
  ripple={true}
>
  <Styled(View)
    current={false}
  >
    <Icon
      name="repeat"
      size={28}
    />
  </Styled(View)>
</Styled(Link)>
`;
